File: src/test/java/com/koddy/server/coffeechat/domain/repository/query/CoffeeChatScheduleQueryRepositoryFetchMentorCoffeeChatSchedulesTest.java
package com.koddy.server.coffeechat.domain.repository.query;

import com.koddy.server.coffeechat.domain.model.CoffeeChat;
import com.koddy.server.coffeechat.domain.repository.CoffeeChatRepository;
import com.koddy.server.coffeechat.domain.repository.query.response.MentorCoffeeChatScheduleData;
import com.koddy.server.coffeechat.domain.repository.query.spec.MentorCoffeeChatQueryCondition;
import com.koddy.server.common.fixture.CoffeeChatFixture.MenteeFlow;
import com.koddy.server.common.fixture.CoffeeChatFixture.MentorFlow;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Import;
import org.springframework.data.domain.Slice;

import java.util.List;

import static com.koddy.server.coffeechat.domain.model.CoffeeChatStatus.MENTEE_APPLY;
import static com.koddy.server.coffeechat.domain.model.CoffeeChatStatus.MENTEE_APPLY_COFFEE_CHAT_COMPLETE;
import static com.koddy.server.coffeechat.domain.model.CoffeeChatStatus.MENTEE_CANCEL;
import static com.koddy.server.coffeechat.domain.model.CoffeeChatStatus.MENTEE_PENDING;
import static com.koddy.server.coffeechat.domain.model.CoffeeChatStatus.MENTEE_REJECT;
import static com.koddy.server.coffeechat.domain.model.CoffeeChatStatus.MENTOR_APPROVE;
import static com.koddy.server.coffeechat.domain.model.CoffeeChatStatus.MENTOR_CANCEL;
import static com.koddy.server.coffeechat.domain.model.CoffeeChatStatus.MENTOR_FINALLY_APPROVE;
import static com.koddy.server.coffeechat.domain.model.CoffeeChatStatus.MENTOR_FINALLY_REJECT;
import static com.koddy.server.coffeechat.domain.model.CoffeeChatStatus.MENTOR_REJECT;
import static com.koddy.server.coffeechat.domain.model.CoffeeChatStatus.MENTOR_SUGGEST;
import static com.koddy.server.coffeechat.domain.model.CoffeeChatStatus.MENTOR_SUGGEST_COFFEE_CHAT_COMPLETE;
import static com.koddy.server.common.fixture.CoffeeChatFixture.금요일_1주차_20_00_시작;
import static com.koddy.server.common.fixture.CoffeeChatFixture.금요일_2주차_20_00_시작;
import static com.koddy.server.common.fixture.CoffeeChatFixture.금요일_3주차_20_00_시작;
import static com.koddy.server.common.fixture.CoffeeChatFixture.수요일_1주차_20_00_시작;
import static com.koddy.server.common.fixture.CoffeeChatFixture.수요일_2주차_20_00_시작;
import static com.koddy.server.common.fixture.CoffeeChatFixture.수요일_3주차_20_00_시작;
import static com.koddy.server.common.fixture.CoffeeChatFixture.월요일_1주차_20_00_시작;
import static com.koddy.server.common.fixture.CoffeeChatFixture.월요일_2주차_20_00_시작;
import static com.koddy.server.common.fixture.CoffeeChatFixture.월요일_3주차_20_00_시작;
import static com.koddy.server.common.fixture.CoffeeChatFixture.토요일_1주차_20_00_시작;
import static com.koddy.server.common.fixture.CoffeeChatFixture.토요일_2주차_20_00_시작;
import static com.koddy.server.common.fixture.CoffeeChatFixture.토요일_3주차_20_00_시작;
import static com.koddy.server.common.fixture.CoffeeChatFixture.화요일_1주차_20_00_시작;
import static com.koddy.server.common.fixture.CoffeeChatFixture.화요일_2주차_20_00_시작;
import static com.koddy.server.common.fixture.CoffeeChatFixture.화요일_3주차_20_00_시작;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;

@Import(CoffeeChatScheduleQueryRepositoryImpl.class)
@DisplayName("CoffeeChat -> CoffeeChatScheduleQueryRepository [fetchMentorCoffeeChatSchedules] 테스트")
public class CoffeeChatScheduleQueryRepositoryFetchMentorCoffeeChatSchedulesTest extends CoffeeChatScheduleQueryRepositorySupporter {
    @Autowired
    private CoffeeChatScheduleQueryRepositoryImpl sut;

    @Autowired
    private CoffeeChatRepository coffeeChatRepository;

    private CoffeeChat[] coffeeChats = new CoffeeChat[30];

    @BeforeEach
    void setUp() {
        initMembers();
        coffeeChats = coffeeChatRepository.saveAll(List.of(
                MenteeFlow.apply(월요일_1주차_20_00_시작, mentees[0], mentors[0]), // 대기
                MenteeFlow.applyAndApprove(화요일_1주차_20_00_시작, mentees[1], mentors[0]), // 예정
                MenteeFlow.apply(수요일_1주차_20_00_시작, mentees[2], mentors[0]), // 대기
                MenteeFlow.applyAndReject(토요일_1주차_20_00_시작, mentees[3], mentors[0]), // 지나간
                MenteeFlow.applyAndCancel(금요일_1주차_20_00_시작, mentees[4], mentors[0]), // 지나간
                MenteeFlow.applyAndApprove(월요일_2주차_20_00_시작, mentees[5], mentors[0]), // 예정
                MenteeFlow.apply(화요일_2주차_20_00_시작, mentees[6], mentors[0]), // 대기
                MenteeFlow.applyAndApprove(수요일_2주차_20_00_시작, mentees[7], mentors[0]), // 예정
                MenteeFlow.applyAndComplete(토요일_2주차_20_00_시작, mentees[8], mentors[0]), // 지나간
                MenteeFlow.apply(금요일_2주차_20_00_시작, mentees[9], mentors[0]), // 대기
                MenteeFlow.applyAndReject(월요일_3주차_20_00_시작, mentees[10], mentors[0]), // 지나간
                MenteeFlow.apply(화요일_3주차_20_00_시작, mentees[11], mentors[0]), // 대기
                MenteeFlow.applyAndCancel(수요일_3주차_20_00_시작, mentees[12], mentors[0]), // 지나간
                MenteeFlow.applyAndComplete(토요일_3주차_20_00_시작, mentees[13], mentors[0]), // 지나간
                MenteeFlow.applyAndApprove(금요일_3주차_20_00_시작, mentees[14], mentors[0]), // 예정

                MentorFlow.suggest(mentors[0], mentees[15]), // 대기
                MentorFlow.suggestAndFinallyApprove(화요일_1주차_20_00_시작, mentors[0], mentees[16]), // 예정
                MentorFlow.suggest(mentors[0], mentees[17]), // 대기
                MentorFlow.suggestAndCancel(mentors[0], mentees[18]), // 지나간
                MentorFlow.suggest(mentors[0], mentees[19]), // 대기
                MentorFlow.suggestAndFinallyApprove(월요일_2주차_20_00_시작, mentors[0], mentees[0]), // 예정
                MentorFlow.suggestAndPending(화요일_2주차_20_00_시작, mentors[0], mentees[1]), // 대기
                MentorFlow.suggest(mentors[0], mentees[2]), // 대기
                MentorFlow.suggestAndFinallyReject(토요일_2주차_20_00_시작, mentors[0], mentees[3]), // 지나간
                MentorFlow.suggest(mentors[0], mentees[4]), // 대기
                MentorFlow.suggestAndPending(월요일_3주차_20_00_시작, mentors[0], mentees[5]), // 대기
                MentorFlow.suggest(mentors[0], mentees[6]), // 대기
                MentorFlow.suggestAndComplete(수요일_3주차_20_00_시작, mentors[0], mentees[7]), // 지나간
                MentorFlow.suggest(mentors[0], mentees[8]), // 대기
                MentorFlow.suggestAndPending(토요일_3주차_20_00_시작, mentors[0], mentees[9]) // 대기
        )).toArray(CoffeeChat[]::new);
    }

    @Test
    @DisplayName("1. 멘토의 내 일정 `대기 상태` 커피챗 정보를 조회한다 [MENTEE_APPLY / MENTOR_SUGGEST / MENTEE_PENDING]")
    void waiting() {
        // given
        final MentorCoffeeChatQueryCondition condition = new MentorCoffeeChatQueryCondition(
                mentors[0].getId(),
                List.of(MENTEE_APPLY, MENTOR_SUGGEST, MENTEE_PENDING)
        );

        /* 페이지 1 */
        final Slice<MentorCoffeeChatScheduleData> result1 = sut.fetchMentorCoffeeChatSchedules(condition, pageable1);
        assertAll(
                () -> assertThat(result1.hasNext()).isTrue(),
                () -> assertThat(result1.getContent())
                        .map(MentorCoffeeChatScheduleData::id)
                        .containsExactly(
                                coffeeChats[29].getId(), coffeeChats[28].getId(), coffeeChats[26].getId(),
                                coffeeChats[25].getId(), coffeeChats[24].getId(), coffeeChats[22].getId(),
                                coffeeChats[21].getId(), coffeeChats[19].getId(), coffeeChats[17].getId(),
                                coffeeChats[15].getId()
                        ),
                () -> assertThat(result1.getContent())
                        .map(MentorCoffeeChatScheduleData::status)
                        .containsExactly(
                                coffeeChats[29].getStatus().name(), coffeeChats[28].getStatus().name(), coffeeChats[26].getStatus().name(),
                                coffeeChats[25].getStatus().name(), coffeeChats[24].getStatus().name(), coffeeChats[22].getStatus().name(),
                                coffeeChats[21].getStatus().name(), coffeeChats[19].getStatus().name(), coffeeChats[17].getStatus().name(),
                                coffeeChats[15].getStatus().name()
                        ),
                () -> assertThat(result1.getContent())
                        .map(MentorCoffeeChatScheduleData::menteeId)
                        .containsExactly(
                                mentees[9].getId(), mentees[8].getId(), mentees[6].getId(),
                                mentees[5].getId(), mentees[4].getId(), mentees[2].getId(),
                                mentees[1].getId(), mentees[19].getId(), mentees[17].getId(),
                                mentees[15].getId()
                        )
        );

        /* 페이지 2 */
        final Slice<MentorCoffeeChatScheduleData> result2 = sut.fetchMentorCoffeeChatSchedules(condition, pageable2);
        assertAll(
                () -> assertThat(result2.hasNext()).isFalse(),
                () -> assertThat(result2.getContent())
                        .map(MentorCoffeeChatScheduleData::id)
                        .containsExactly(
                                coffeeChats[11].getId(), coffeeChats[9].getId(), coffeeChats[6].getId(),
                                coffeeChats[2].getId(), coffeeChats[0].getId()
                        ),
                () -> assertThat(result2.getContent())
                        .map(MentorCoffeeChatScheduleData::status)
                        .containsExactly(
                                coffeeChats[11].getStatus().name(), coffeeChats[9].getStatus().name(), coffeeChats[6].getStatus().name(),
                                coffeeChats[2].getStatus().name(), coffeeChats[0].getStatus().name()
                        ),
                () -> assertThat(result2.getContent())
                        .map(MentorCoffeeChatScheduleData::menteeId)
                        .containsExactly(
                                mentees[11].getId(), mentees[9].getId(), mentees[6].getId(),
                                mentees[2].getId(), mentees[0].getId()
                        )
        );
    }

    @Test
    @DisplayName("2. 멘토의 내 일정 `예정 상태` 커피챗 정보를 조회한다 [MENTOR_APPROVE & MENTOR_FINALLY_APPROVE]")
    void scheduled() {
        // given
        final MentorCoffeeChatQueryCondition condition = new MentorCoffeeChatQueryCondition(
                mentors[0].getId(),
                List.of(MENTOR_APPROVE, MENTOR_FINALLY_APPROVE)
        );

        /* 페이지 1 */
        final Slice<MentorCoffeeChatScheduleData> result1 = sut.fetchMentorCoffeeChatSchedules(condition, pageable1);
        assertAll(
                () -> assertThat(result1.hasNext()).isFalse(),
                () -> assertThat(result1.getContent())
                        .map(MentorCoffeeChatScheduleData::id)
                        .containsExactly(
                                coffeeChats[20].getId(), coffeeChats[16].getId(), coffeeChats[14].getId(),
                                coffeeChats[7].getId(), coffeeChats[5].getId(), coffeeChats[1].getId()
                        ),
                () -> assertThat(result1.getContent())
                        .map(MentorCoffeeChatScheduleData::status)
                        .containsExactly(
                                coffeeChats[20].getStatus().name(), coffeeChats[16].getStatus().name(), coffeeChats[14].getStatus().name(),
                                coffeeChats[7].getStatus().name(), coffeeChats[5].getStatus().name(), coffeeChats[1].getStatus().name()
                        ),
                () -> assertThat(result1.getContent())
                        .map(MentorCoffeeChatScheduleData::menteeId)
                        .containsExactly(
                                mentees[0].getId(), mentees[16].getId(), mentees[14].getId(),
                                mentees[7].getId(), mentees[5].getId(), mentees[1].getId()
                        )
        );

        /* 페이지 2 */
        final Slice<MentorCoffeeChatScheduleData> result2 = sut.fetchMentorCoffeeChatSchedules(condition, pageable2);
        assertAll(
                () -> assertThat(result2.hasNext()).isFalse(),
                () -> assertThat(result2.getContent()).isEmpty()
        );
    }

    @Test
    @DisplayName("3. 멘토의 내 일정 `지나간 상태` 커피챗 정보를 조회한다 [MENTEE_CANCEL / MENTOR_REJECT / MENTEE_APPLY_COFFEE_CHAT_COMPLETE / MENTOR_CANCEL / MENTEE_REJECT / MENTOR_FINALLY_REJECT / MENTOR_SUGGEST_COFFEE_CHAT_COMPLETE]")
    void passed() {
        // given
        final MentorCoffeeChatQueryCondition condition = new MentorCoffeeChatQueryCondition(
                mentors[0].getId(),
                List.of(
                        MENTEE_CANCEL, MENTOR_REJECT, MENTEE_APPLY_COFFEE_CHAT_COMPLETE,
                        MENTOR_CANCEL, MENTEE_REJECT, MENTOR_FINALLY_REJECT, MENTOR_SUGGEST_COFFEE_CHAT_COMPLETE
                )
        );

        /* 페이지 1 */
        final Slice<MentorCoffeeChatScheduleData> result1 = sut.fetchMentorCoffeeChatSchedules(condition, pageable1);
        assertAll(
                () -> assertThat(result1.hasNext()).isFalse(),
                () -> assertThat(result1.getContent())
                        .map(MentorCoffeeChatScheduleData::id)
                        .containsExactly(
                                coffeeChats[27].getId(), coffeeChats[23].getId(), coffeeChats[18].getId(),
                                coffeeChats[13].getId(), coffeeChats[12].getId(), coffeeChats[10].getId(),
                                coffeeChats[8].getId(), coffeeChats[4].getId(), coffeeChats[3].getId()
                        ),
                () -> assertThat(result1.getContent())
                        .map(MentorCoffeeChatScheduleData::status)
                        .containsExactly(
                                coffeeChats[27].getStatus().name(), coffeeChats[23].getStatus().name(), coffeeChats[18].getStatus().name(),
                                coffeeChats[13].getStatus().name(), coffeeChats[12].getStatus().name(), coffeeChats[10].getStatus().name(),
                                coffeeChats[8].getStatus().name(), coffeeChats[4].getStatus().name(), coffeeChats[3].getStatus().name()
                        ),
                () -> assertThat(result1.getContent())
                        .map(MentorCoffeeChatScheduleData::menteeId)
                        .containsExactly(
                                mentees[7].getId(), mentees[3].getId(), mentees[18].getId(),
                                mentees[13].getId(), mentees[12].getId(), mentees[10].getId(),
                                mentees[8].getId(), mentees[4].getId(), mentees[3].getId()
                        )
        );

        /* 페이지 2 */
        final Slice<MentorCoffeeChatScheduleData> result2 = sut.fetchMentorCoffeeChatSchedules(condition, pageable2);
        assertAll(
                () -> assertThat(result2.hasNext()).isFalse(),
                () -> assertThat(result2.getContent()).isEmpty()
        );
    }
}

File: src/test/java/com/koddy/server/coffeechat/domain/repository/query/CoffeeChatScheduleQueryRepositoryFetchMentorCoffeeChatSchedulesTest.kt
package com.koddy.server.coffeechat.domain.repository.query

import com.koddy.server.coffeechat.domain.model.CoffeeChat
import com.koddy.server.coffeechat.domain.model.CoffeeChatStatus
import com.koddy.server.coffeechat.domain.repository.CoffeeChatRepository
import com.koddy.server.coffeechat.domain.repository.query.response.CoffeeChatCountPerCategory
import com.koddy.server.coffeechat.domain.repository.query.response.MentorCoffeeChatScheduleData
import com.koddy.server.coffeechat.domain.repository.query.spec.MentorCoffeeChatQueryCondition
import com.koddy.server.common.RepositoryTestKt
import com.koddy.server.common.fixture.CoffeeChatFixture.금요일_1주차_20_00_시작
import com.koddy.server.common.fixture.CoffeeChatFixture.금요일_2주차_20_00_시작
import com.koddy.server.common.fixture.CoffeeChatFixture.금요일_3주차_20_00_시작
import com.koddy.server.common.fixture.CoffeeChatFixture.수요일_1주차_20_00_시작
import com.koddy.server.common.fixture.CoffeeChatFixture.수요일_2주차_20_00_시작
import com.koddy.server.common.fixture.CoffeeChatFixture.수요일_3주차_20_00_시작
import com.koddy.server.common.fixture.CoffeeChatFixture.월요일_1주차_20_00_시작
import com.koddy.server.common.fixture.CoffeeChatFixture.월요일_2주차_20_00_시작
import com.koddy.server.common.fixture.CoffeeChatFixture.월요일_3주차_20_00_시작
import com.koddy.server.common.fixture.CoffeeChatFixture.토요일_1주차_20_00_시작
import com.koddy.server.common.fixture.CoffeeChatFixture.토요일_2주차_20_00_시작
import com.koddy.server.common.fixture.CoffeeChatFixture.토요일_3주차_20_00_시작
import com.koddy.server.common.fixture.CoffeeChatFixture.화요일_1주차_20_00_시작
import com.koddy.server.common.fixture.CoffeeChatFixture.화요일_2주차_20_00_시작
import com.koddy.server.common.fixture.CoffeeChatFixture.화요일_3주차_20_00_시작
import com.koddy.server.common.fixture.MenteeFlow
import com.koddy.server.common.fixture.MentorFlow
import com.koddy.server.member.domain.repository.MemberRepository
import io.kotest.assertions.assertSoftly
import io.kotest.matchers.collections.shouldContainExactly
import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.springframework.context.annotation.Import
import org.springframework.data.domain.Slice

@RepositoryTestKt
@Import(CoffeeChatScheduleQueryRepositoryImpl::class)
@DisplayName("CoffeeChat -> CoffeeChatScheduleQueryRepository [fetchMentorCoffeeChatSchedules] 테스트")
class CoffeeChatScheduleQueryRepositoryFetchMentorCoffeeChatSchedulesTest(
    private val sut: CoffeeChatScheduleQueryRepositoryImpl,
    private val coffeeChatRepository: CoffeeChatRepository,
    memberRepository: MemberRepository,
) : CoffeeChatScheduleQueryRepositorySupporter(memberRepository) {
    private lateinit var coffeeChats: List<CoffeeChat>

    @BeforeEach
    fun setUp() {
        initMembers()
        coffeeChats = coffeeChatRepository.saveAll(
            listOf(
                MenteeFlow.apply(fixture = 월요일_1주차_20_00_시작, mentee = mentees[0], mentor = mentors[0]),  // 대기
                MenteeFlow.applyAndApprove(fixture = 화요일_1주차_20_00_시작, mentee = mentees[1], mentor = mentors[0]),  // 예정
                MenteeFlow.apply(fixture = 수요일_1주차_20_00_시작, mentee = mentees[2], mentor = mentors[0]),  // 대기
                MenteeFlow.applyAndReject(fixture = 토요일_1주차_20_00_시작, mentee = mentees[3], mentor = mentors[0]),  // 지나간
                MenteeFlow.applyAndCancel(fixture = 금요일_1주차_20_00_시작, mentee = mentees[4], mentor = mentors[0]),  // 지나간
                MenteeFlow.apply(fixture = 월요일_2주차_20_00_시작, mentee = mentees[5], mentor = mentors[0]),  // 대기
                MenteeFlow.apply(fixture = 화요일_2주차_20_00_시작, mentee = mentees[6], mentor = mentors[0]),  // 대기
                MenteeFlow.applyAndApprove(fixture = 수요일_2주차_20_00_시작, mentee = mentees[7], mentor = mentors[0]),  // 예정
                MenteeFlow.applyAndComplete(fixture = 토요일_2주차_20_00_시작, mentee = mentees[8], mentor = mentors[0]),  // 지나간
                MenteeFlow.apply(fixture = 금요일_2주차_20_00_시작, mentee = mentees[9], mentor = mentors[0]),  // 대기
                MenteeFlow.applyAndReject(fixture = 월요일_3주차_20_00_시작, mentee = mentees[10], mentor = mentors[0]),  // 지나간
                MenteeFlow.apply(fixture = 화요일_3주차_20_00_시작, mentee = mentees[11], mentor = mentors[0]),  // 대기
                MenteeFlow.applyAndCancel(fixture = 수요일_3주차_20_00_시작, mentee = mentees[12], mentor = mentors[0]),  // 지나간
                MenteeFlow.applyAndComplete(fixture = 토요일_3주차_20_00_시작, mentee = mentees[13], mentor = mentors[0]),  // 지나간
                MenteeFlow.apply(fixture = 금요일_3주차_20_00_시작, mentee = mentees[14], mentor = mentors[0]),  // 대기

                MentorFlow.suggest(mentor = mentors[0], mentee = mentees[15]),  // 제안
                MentorFlow.suggestAndPending(fixture = 화요일_1주차_20_00_시작, mentor = mentors[0], mentee = mentees[16]),  // 대기
                MentorFlow.suggest(mentor = mentors[0], mentee = mentees[17]),  // 제안
                MentorFlow.suggestAndCancel(mentor = mentors[0], mentee = mentees[18]),  // 지나간
                MentorFlow.suggest(mentor = mentors[0], mentee = mentees[19]),  // 제안
                MentorFlow.suggestAndFinallyApprove(fixture = 월요일_2주차_20_00_시작, mentor = mentors[0], mentee = mentees[0]),  // 예정
                MentorFlow.suggestAndPending(fixture = 화요일_2주차_20_00_시작, mentor = mentors[0], mentee = mentees[1]),  // 대기
                MentorFlow.suggest(mentor = mentors[0], mentee = mentees[2]),  // 제안
                MentorFlow.suggestAndFinallyCancel(fixture = 토요일_2주차_20_00_시작, mentor = mentors[0], mentee = mentees[3]),  // 지나간
                MentorFlow.suggest(mentor = mentors[0], mentee = mentees[4]),  // 제안
                MentorFlow.suggestAndPending(fixture = 월요일_3주차_20_00_시작, mentor = mentors[0], mentee = mentees[5]),  // 대기
                MentorFlow.suggest(mentor = mentors[0], mentee = mentees[6]),  // 제안
                MentorFlow.suggestAndComplete(fixture = 수요일_3주차_20_00_시작, mentor = mentors[0], mentee = mentees[7]),  // 지나간
                MentorFlow.suggest(mentor = mentors[0], mentee = mentees[8]),  // 제안
                MentorFlow.suggestAndPending(fixture = 토요일_3주차_20_00_시작, mentor = mentors[0], mentee = mentees[9]), // 대기
            ),
        )
    }

    @Test
    @DisplayName("0. 멘토의 상태별 커피챗 개수를 조회한다 [대기, 제안, 예정, 지나간]")
    fun counts() {
        val result: CoffeeChatCountPerCategory = sut.fetchMentorCoffeeChatCountPerCategory(mentors[0].id)
        assertSoftly {
            result.waiting shouldBe 11
            coffeeChatRepository.countByMentorIdAndStatusIn(mentors[0].id, CoffeeChatStatus.withWaitingCategory()) shouldBe 11
            result.suggested shouldBe 7
            coffeeChatRepository.countByMentorIdAndStatusIn(mentors[0].id, CoffeeChatStatus.withSuggstedCategory()) shouldBe 7
            result.scheduled shouldBe 3
            coffeeChatRepository.countByMentorIdAndStatusIn(mentors[0].id, CoffeeChatStatus.withScheduledCategory()) shouldBe 3
            result.passed shouldBe 9
            coffeeChatRepository.countByMentorIdAndStatusIn(mentors[0].id, CoffeeChatStatus.withPassedCategory()) shouldBe 9
        }
    }

    @Test
    @DisplayName("1. 멘토의 내 일정 `대기 상태` 커피챗 정보를 조회한다")
    fun waiting() {
        // given
        val condition = MentorCoffeeChatQueryCondition(mentors[0].id, CoffeeChatStatus.withWaitingCategory())

        /* 페이지 1 */
        val result1: Slice<MentorCoffeeChatScheduleData> = sut.fetchMentorCoffeeChatSchedules(condition, pageable1)
        assertSoftly(result1) {
            hasNext() shouldBe true
            content.map { it.id } shouldContainExactly listOf(
                coffeeChats[29].id, coffeeChats[25].id, coffeeChats[21].id,
                coffeeChats[16].id, coffeeChats[14].id, coffeeChats[11].id,
                coffeeChats[9].id, coffeeChats[6].id, coffeeChats[5].id,
                coffeeChats[2].id,
            )
            content.map { it.status } shouldContainExactly listOf(
                coffeeChats[29].status, coffeeChats[25].status, coffeeChats[21].status,
                coffeeChats[16].status, coffeeChats[14].status, coffeeChats[11].status,
                coffeeChats[9].status, coffeeChats[6].status, coffeeChats[5].status,
                coffeeChats[2].status,
            )
            content.map { it.menteeId } shouldContainExactly listOf(
                mentees[9].id, mentees[5].id, mentees[1].id,
                mentees[16].id, mentees[14].id, mentees[11].id,
                mentees[9].id, mentees[6].id, mentees[5].id,
                mentees[2].id,
            )
        }

        /* 페이지 2 */
        val result2: Slice<MentorCoffeeChatScheduleData> = sut.fetchMentorCoffeeChatSchedules(condition, pageable2)
        assertSoftly(result2) {
            hasNext() shouldBe false
            content.map { it.id } shouldContainExactly listOf(coffeeChats[0].id)
            content.map { it.status } shouldContainExactly listOf(coffeeChats[0].status)
            content.map { it.menteeId } shouldContainExactly listOf(mentees[0].id)
        }
    }

    @Test
    @DisplayName("2. 멘토의 내 일정 `제안 상태` 커피챗 정보를 조회한다")
    fun suggest() {
        // given
        val condition = MentorCoffeeChatQueryCondition(mentors[0].id, CoffeeChatStatus.withSuggstedCategory())

        /* 페이지 1 */
        val result1: Slice<MentorCoffeeChatScheduleData> = sut.fetchMentorCoffeeChatSchedules(condition, pageable1)
        assertSoftly(result1) {
            hasNext() shouldBe false
            content.map { it.id } shouldContainExactly listOf(
                coffeeChats[28].id, coffeeChats[26].id, coffeeChats[24].id,
                coffeeChats[22].id, coffeeChats[19].id, coffeeChats[17].id,
                coffeeChats[15].id,
            )
            content.map { it.status } shouldContainExactly listOf(
                coffeeChats[28].status, coffeeChats[26].status, coffeeChats[24].status,
                coffeeChats[22].status, coffeeChats[19].status, coffeeChats[17].status,
                coffeeChats[15].status,
            )
            content.map { it.menteeId } shouldContainExactly listOf(
                mentees[8].id, mentees[6].id, mentees[4].id,
                mentees[2].id, mentees[19].id, mentees[17].id,
                mentees[15].id,
            )
        }

        /* 페이지 2 */
        val result2: Slice<MentorCoffeeChatScheduleData> = sut.fetchMentorCoffeeChatSchedules(condition, pageable2)
        assertSoftly(result2) {
            hasNext() shouldBe false
            content shouldBe emptyList()
        }
    }

    @Test
    @DisplayName("3. 멘토의 내 일정 `예정 상태` 커피챗 정보를 조회한다")
    fun scheduled() {
        // given
        val condition = MentorCoffeeChatQueryCondition(mentors[0].id, CoffeeChatStatus.withScheduledCategory())

        /* 페이지 1 */
        val result1: Slice<MentorCoffeeChatScheduleData> = sut.fetchMentorCoffeeChatSchedules(condition, pageable1)
        assertSoftly(result1) {
            hasNext() shouldBe false
            content.map { it.id } shouldContainExactly listOf(coffeeChats[20].id, coffeeChats[7].id, coffeeChats[1].id)
            content.map { it.status } shouldContainExactly listOf(coffeeChats[20].status, coffeeChats[7].status, coffeeChats[1].status)
            content.map { it.menteeId } shouldContainExactly listOf(mentees[0].id, mentees[7].id, mentees[1].id)
        }

        /* 페이지 2 */
        val result2: Slice<MentorCoffeeChatScheduleData> = sut.fetchMentorCoffeeChatSchedules(condition, pageable2)
        assertSoftly(result2) {
            hasNext() shouldBe false
            content shouldBe emptyList()
        }
    }

    @Test
    @DisplayName("4. 멘토의 내 일정 `지나간 상태` 커피챗 정보를 조회한다")
    fun passed() {
        // given
        val condition = MentorCoffeeChatQueryCondition(mentors[0].id, CoffeeChatStatus.withPassedCategory())

        /* 페이지 1 */
        val result1: Slice<MentorCoffeeChatScheduleData> = sut.fetchMentorCoffeeChatSchedules(condition, pageable1)
        assertSoftly(result1) {
            hasNext() shouldBe false
            content.map { it.id } shouldContainExactly listOf(
                coffeeChats[27].id, coffeeChats[23].id, coffeeChats[18].id,
                coffeeChats[13].id, coffeeChats[12].id, coffeeChats[10].id,
                coffeeChats[8].id, coffeeChats[4].id, coffeeChats[3].id,
            )
            content.map { it.status } shouldContainExactly listOf(
                coffeeChats[27].status, coffeeChats[23].status, coffeeChats[18].status,
                coffeeChats[13].status, coffeeChats[12].status, coffeeChats[10].status,
                coffeeChats[8].status, coffeeChats[4].status, coffeeChats[3].status,
            )
            content.map { it.menteeId } shouldContainExactly listOf(
                mentees[7].id, mentees[3].id, mentees[18].id,
                mentees[13].id, mentees[12].id, mentees[10].id,
                mentees[8].id, mentees[4].id, mentees[3].id,
            )
        }

        /* 페이지 2 */
        val result2: Slice<MentorCoffeeChatScheduleData> = sut.fetchMentorCoffeeChatSchedules(condition, pageable2)
        assertSoftly(result2) {
            hasNext() shouldBe false
            content shouldBe emptyList()
        }
    }
}
